### PR TITLE
Add DateTime HIL, and implement for RPi Pico and STM32f429ZI

### PR DESCRIPTION
**Pull request by Remus7**
### Pull Request Overview

This pull request adds a Date and Time HIL, driver and component. This includes low level drivers for Real Time Clock for nucleo-stm32f429zi and raspberry pi pico boards. The Real Time Clock for raspberry pi pico board was rebased for an older commit written for tock 1.0 (https://github.com/tock/tock/pull/2760). The code compiles with other changes made to the libtock-c project, including an example project to test Real Time Clock intended usage (https://github.com/tock/libtock-c/pull/334).

### Testing Strategy

This pull request was tested by running the implemented commands both in the board/main.src for the nucleo board, and also by running a program from the libtock-c userspace (setting and then printing current time each second).

### TODO or Help Wanted

The Real Time Clock works for the two boards, RTC support for more boards might be added using the component, hil and capsule provided.

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make prepush`.
- [x] Ran `make format`.

### CODE REVIEW

File: kernel/src/hil/date_time.rs [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    /// Sets a client that calls the callback function when date and time is requested
+    fn set_client(&self, client: &'a dyn DateTimeClient);
+}
+
+/// Callback handler for when current date is read or set.
+pub trait DateTimeClient {
+    /// Called when a date time reading has completed.
+    /// Takes `Ok(DateTime)` of current date and passes it when scheduling an upcall.
+    /// If an error is encountered it takes an `Err(ErrorCode)`
+    fn callback_get_date(&self, datetime: Result<Date, ErrorCode>);
```
Comment (bradjc) [inline kernel/src/hil/date_time.rs:71]: `callback` is redundant here, what about `got_date_time()` or `current_date_time()`?

File: kernel/src/hil/date_time.rs [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+    /// Called when a date is set
+    /// Takes `Ok(())` if time is set correctly.
+    /// Takes  `Err(ErrorCode)` in case of an error
+    fn callback_set_date(&self, result: Result<(), ErrorCode>);
```
Comment (bradjc) [inline kernel/src/hil/date_time.rs:76]:
```suggestion
    fn set_date_time_done(&self, result: Result<(), ErrorCode>);
```
would be more consistent with other HILs.

File: kernel/src/hil/date_time.rs [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    pub seconds: u8,
+}
+
+/// Interface for reading and setting the current time
+pub trait DateTime<'a> {
+    /// Request driver to return date and time
+    ///
+    /// When successful this function call must be followed by a call
+    /// to `callback_get_date` which provides the actual date and time
+    /// or an error.
```
Comment (bradjc) [inline kernel/src/hil/date_time.rs:53]:
```suggestion
    /// When successful, this function will be followed by the callback
    /// `callback_get_date` which provides the actual date and time
    /// or an error.
```

File: kernel/src/hil/date_time.rs [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    July,
+    August,
+    September,
+    October,
+    November,
+    December,
+}
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub struct Date {
```
Comment (bradjc) [inline kernel/src/hil/date_time.rs:37]:
```suggestion
pub struct DateTime {
```
?

File: kernel/src/hil/date_time.rs [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    pub year: u16,
+    pub month: Month,
+    pub day: u8,
+    pub day_of_week: DayOfWeek,
```
Comment (bradjc) [inline kernel/src/hil/date_time.rs:41]: This might be hard for the caller to provide when setting the datetime?

File: chips/stm32f4xx/src/lib.rs [hunk 6]
```diff
@@ -29,6 +29,8 @@ pub mod tim2;
 pub mod trng;
 pub mod usart;
 
+pub mod pwr;
```
Comment (bradjc) [inline chips/stm32f4xx/src/lib.rs:32]: Can you move the stm changes to a different PR? Or are these not supposed to be here?

File: capsules/extra/src/date_time.rs [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+    ///         first number (year, month, day_of_the_month):
+    ///                 -last 5 bits store the day_of_the_month
+    ///                 -previous 4 bits store the month
+    ///                 -previous 12 bits store the year
+    ///         second number (day_of_the_week, hour, minute, seconds):
+    ///                 -last 6 bits store the seconds
+    ///                 -previous 6 store the minute
+    ///                 -previous 5 store the hour
+    ///                 -previous 3 store the day_of_the_week
+
```
Comment (bradjc) [inline capsules/extra/src/date_time.rs:135]:
```suggestion
```

File: capsules/extra/src/date_time.rs [hunk 8]
```diff
@@ -0,0 +1,354 @@
...
+    ///                 -previous 4 bits store the month
+    ///                 -previous 12 bits store the year
+    ///         second number (day_of_the_week, hour, minute, seconds):
+    ///                 -last 6 bits store the seconds
+    ///                 -previous 6 store the minute
+    ///                 -previous 5 store the hour
+    ///                 -previous 3 store the day_of_the_week
+    ///
+    /// The two u32 numbers are returned as a tuple
+
```
Comment (bradjc) [inline capsules/extra/src/date_time.rs:182]:
```suggestion
```

File: capsules/extra/src/date_time.rs [hunk 9]
```diff
@@ -0,0 +1,354 @@
...
+    /// Transforms Date structure (year, month, dotm, dotw, hour, minute, seconds) into two u32 numbers
+    ///         first number (year, month, day_of_the_month):
+    ///                 -last 5 bits store the day_of_the_month
+    ///                 -previous 4 bits store the month
+    ///                 -previous 12 bits store the year
+    ///         second number (day_of_the_week, hour, minute, seconds):
+    ///                 -last 6 bits store the seconds
+    ///                 -previous 6 store the minute
+    ///                 -previous 5 store the hour
+    ///                 -previous 3 store the day_of_the_week
```
Comment (bradjc) [inline capsules/extra/src/date_time.rs:179]: I think this should be moved to the top comment in the file since it's very important to how this driver works.

File: capsules/extra/src/date_time.rs [hunk 10]
```diff
@@ -0,0 +1,354 @@
...
+                        dotw_hour_min_sec as usize,
+                    )
+                    .into(),
+                Err(_e) => CommandReturn::failure(ErrorCode::FAIL),
+            }
+        } else {
+            let grant_enter_res = self.apps.enter(appid, |app, _| {
+                app.subscribed = true;
+            });
+
```
Comment (bradjc) [inline capsules/extra/src/date_time.rs:258]: Need to actually store the command here.

File: capsules/extra/src/date_time.rs [hunk 11]
```diff
@@ -0,0 +1,354 @@
...
+{
+    fn callback_get_date(&self, datetime: Result<date_time::Date, ErrorCode>) {
+        self.in_progress.set(false);
+        let mut upcall_status: usize = into_statuscode(Ok(()));
+        let mut upcall_r1: usize = 0;
+        let mut upcall_r2: usize = 0;
+
+        for cntr in self.apps.iter() {
+            cntr.enter(|app, upcalls| {
+                if app.subscribed {
```
Comment (bradjc) [inline capsules/extra/src/date_time.rs:278]: Need to make sure the app actually wanted to read the date (and not set it).

File: capsules/extra/src/date_time.rs [hunk 12]
```diff
@@ -0,0 +1,354 @@
...
+    fn call_driver(&self, command: DateTimeCommand, r2: usize, r3: usize) -> CommandReturn {
+        match command {
+            DateTimeCommand::ReadDateTime => {
+                let date_result = self.date_time.get_date_time();
+                match date_result {
+                    Result::Ok(()) => {
+                        self.in_progress.set(true);
+                        CommandReturn::success()
+                    }
+                    Result::Err(e) => CommandReturn::failure(e),
```
Comment (bradjc) [inline capsules/extra/src/date_time.rs:207]: This potentially needs to generate an upcall. If a read was queued after a set, and when this read happens an error occurs, then the app calling get will never get a callback. But, if nothing was queued, then this can return an error directly in response to the command syscall.

Comment (bradjc) [reply capsules/extra/src/date_time.rs:207]: This is true of all command return errors in this function.

File: capsules/extra/src/date_time.rs [hunk 13]
```diff
@@ -0,0 +1,354 @@
...
+                }
+            });
+        }
+    }
+
+    fn callback_set_date(&self, result: Result<(), ErrorCode>) {
+        self.in_progress.set(false);
+
+        for cntr in self.apps.iter() {
+            //let mut upcall_status = UPCALL_OK;
```
Comment (bradjc) [inline capsules/extra/src/date_time.rs:311]:
```suggestion
```

File: boards/raspberry_pi_pico/src/main.rs [hunk 14]
```diff
@@ -451,6 +459,42 @@ pub unsafe fn main() {
...
+    // };
+    //
+    // match peripherals.rtc.set_date_time(dt) {
+    //     Err(e) => {
+    //         debug!("Error setting date and time {:?}", e);
+    //     }
+    //     _ => {
+    //         debug!("Date and time are set");
+    //     }
+    // };
```
Comment (bradjc) [inline boards/raspberry_pi_pico/src/main.rs:496]: Leaving this here makes me nervous. Since the date time driver is already setup, that driver is going to get the callback when this operation finishes. Now, that is probably benign in practice, but, it's not good practice for code to get callbacks that it didn't request.

I don't think we should include this example in default tock boards because it isn't something we would like replicated.

Now, if there was a virtual_date_time.rs layer, then the kernel could do this safely.

File: kernel/src/hil/date_time.rs (hunk 1 above)
Comment (Remus7) [reply kernel/src/hil/date_time.rs:71]: Can it not be get_date_time_done(), in order to be consistent with the set_date_time_done() function?

File: chips/stm32f4xx/src/lib.rs (hunk 6 above)
Comment (Remus7) [reply chips/stm32f4xx/src/lib.rs:32]: The user manual that I used specified that the PWR registers applied for the whole STM32F4xx family. That being said, I did not have the opportunity to test it on any other boards than the nucleo-stm32f429zi.

[ST_Processor_Manual.pdf](https://github.com/tock/tock/files/12214393/ST_Processor_Manual.pdf)

File: kernel/src/hil/date_time.rs (hunk 4 above)
Comment (Remus7) [reply kernel/src/hil/date_time.rs:37]: The name is supposed to differentiate easily between the DateTime trait (for getting and setting the date) and the actual Date structure getting passed around.

**Comment by Remus7** [comment]
Thank you very much for the suggestions. I updated the function names and the capsule documentation as instructed. 
However, for the next week I will be unable to test the actual code, since I do not have access to any stm32f429zi boards. This is why I would rather wait before making any large changes to the code, so that I am able to test for any unforseen effects on the rest of the code

File: kernel/src/hil/date_time.rs (hunk 1 above)
Comment (bradjc) [reply kernel/src/hil/date_time.rs:71]: Sounds good to me.

File: kernel/src/hil/date_time.rs (hunk 4 above)
Comment (bradjc) [reply kernel/src/hil/date_time.rs:37]: Maybe `DateTimeValues`? This struct contains more than just date information.

File: kernel/src/hil/date_time.rs (hunk 5 above)
Comment (bradjc) [reply kernel/src/hil/date_time.rs:41]: But what if you want to use both date and time, you have to know the day of the week too. And what if an RTC provides date and time, but not day of week? I just wonder if making it an `Option<DayOfWeek>` would be more portable.

File: chips/stm32f4xx/src/lib.rs (hunk 6 above)
Comment (bradjc) [reply chips/stm32f4xx/src/lib.rs:32]: I thought this PR was just for the RPi Pico RTC?

File: kernel/src/hil/date_time.rs (hunk 4 above)
Comment (Remus7) [reply kernel/src/hil/date_time.rs:37]: That is reasonable. I will rename the struct.

File: kernel/src/hil/date_time.rs (hunk 5 above)
Comment (Remus7) [reply kernel/src/hil/date_time.rs:41]: From what I understood, the board RTC will update the DayOfWeek regardless of whether it was set or not. If the user doesn't need to use the DayOfWeek field he can simply ignore it, but in the back, the board will still update the DayOfWeek registers beginning with the value 0 (which is either Monday, Saturday or null depending on the board). I do not think making it Option<DayOfWeek> has any real advantage.

File: chips/stm32f4xx/src/lib.rs (hunk 6 above)
Comment (Remus7) [reply chips/stm32f4xx/src/lib.rs:32]: This PR also includes RTC support for the nucleo_f429zi board (edited the chips folder for stm32f429 and the board foler for nucleo_f429zi). The code for RPi Pico RTC was mostly rebased from an older commit, while the code for the nucleo board was written by me.
In any case, what should I do with the PWR registers? The code is generic for all the STM32F42xxx and STM32F43xxx boards, so it would be a loss to just use it for the specific nucleo_f429zi board.

File: kernel/src/hil/date_time.rs (hunk 5 above)
Comment (bradjc) [reply kernel/src/hil/date_time.rs:41]: Ok I just googled for "low power rtc" and the first three results from three different companies all support day of the week. Please ignore my original comment.

File: chips/stm32f429zi/src/rtc.rs [hunk 15]
```diff
@@ -0,0 +1,651 @@
...
+//!
+//! + Set time from which real time clock should start counting
+//! + Read current time from the RTC registers
+//!
+
+use core::cell::Cell;
+use kernel::deferred_call::{DeferredCall, DeferredCallClient};
+use kernel::platform::chip::ClockInterface;
+use stm32f4xx::rcc;
+// use kernel::static_init;
```
Comment (bradjc) [inline chips/stm32f429zi/src/rtc.rs:22]:
```suggestion
```

File: chips/stm32f429zi/src/rtc.rs [hunk 16]
```diff
@@ -0,0 +1,651 @@
...
+            Month::July => 7,
+            Month::August => 8,
+            Month::September => 9,
+            Month::October => 10,
+            Month::November => 11,
+            Month::December => 12,
+        }
+    }
+
+    #[inline(never)]
```
Comment (bradjc) [inline chips/stm32f429zi/src/rtc.rs:479]: Can you comment why we need these directives?

File: boards/nucleo_f429zi/openocd.cfg [hunk 17]
```diff
@@ -12,4 +12,3 @@ hla_vid_pid 0x0483 0x374b
 set WORKAREASIZE 0x10000
 
 source [find target/stm32f4x.cfg]
-
```
Comment (bradjc) [inline boards/nucleo_f429zi/openocd.cfg:15]: revert

File: boards/nucleo_f429zi/src/main.rs [hunk 18]
```diff
@@ -79,6 +79,11 @@ struct NucleoF429ZI {
     scheduler: &'static RoundRobinSched<'static>,
     systick: cortexm4::systick::SysTick,
     can: &'static capsules_extra::can::CanCapsule<'static, stm32f429zi::can::Can<'static>>,
+
```
Comment (bradjc) [inline boards/nucleo_f429zi/src/main.rs:82]:
```suggestion
```

File: boards/nucleo_f429zi/src/main.rs [hunk 19]
```diff
@@ -627,6 +657,7 @@ pub unsafe fn main() {
         scheduler,
         systick: cortexm4::systick::SysTick::new(),
         can: can,
+        date_time: date_time,
```
Comment (bradjc) [inline boards/nucleo_f429zi/src/main.rs:660]:
```suggestion
        date_time,
```

File: boards/raspberry_pi_pico/src/main.rs [hunk 20]
```diff
@@ -21,9 +21,11 @@ use components::gpio::GpioComponent;
 use components::led::LedsComponent;
 use enum_primitive::cast::FromPrimitive;
 use kernel::component::Component;
+
```
Comment (bradjc) [inline boards/raspberry_pi_pico/src/main.rs:24]:
```suggestion
```

File: boards/raspberry_pi_pico/src/main.rs [hunk 21]
```diff
@@ -21,9 +21,11 @@ use components::gpio::GpioComponent;
...
 use kernel::debug;
 use kernel::hil::gpio::{Configure, FloatingState};
 use kernel::hil::i2c::I2CMaster;
+
```
Comment (bradjc) [inline boards/raspberry_pi_pico/src/main.rs:28]:
```suggestion
```

File: boards/raspberry_pi_pico/src/main.rs [hunk 22]
```diff
@@ -45,6 +47,8 @@ use rp2040::resets::Peripheral;
 use rp2040::sysinfo;
 use rp2040::timer::RPTimer;
 
+use components::date_time_component_static;
+
```
Comment (bradjc) [inline boards/raspberry_pi_pico/src/main.rs:51]:
```suggestion
```

File: boards/raspberry_pi_pico/src/main.rs [hunk 23]
```diff
@@ -96,6 +100,9 @@ pub struct RaspberryPiPico {
     temperature: &'static capsules_extra::temperature::TemperatureSensor<'static>,
     i2c: &'static capsules_core::i2c_master::I2CMasterDriver<'static, I2c<'static, 'static>>,
 
+    date_time:
+        &'static capsules_extra::date_time::DateTimeCapsule<'static, rp2040::rtc::Rtc<'static>>,
+
```
Comment (bradjc) [inline boards/raspberry_pi_pico/src/main.rs:105]:
```suggestion
```

File: boards/raspberry_pi_pico/src/main.rs [hunk 24]
```diff
@@ -481,6 +505,7 @@ pub unsafe fn main() {
                 adc_channel_2,
                 adc_channel_3,
             ));
+
```
Comment (bradjc) [inline boards/raspberry_pi_pico/src/main.rs:508]:
```suggestion
```

File: boards/raspberry_pi_pico/src/main.rs [hunk 25]
```diff
@@ -540,6 +565,7 @@ pub unsafe fn main() {
         adc: adc_syscall,
         temperature: temp,
         i2c,
+        date_time: date_time,
```
Comment (bradjc) [inline boards/raspberry_pi_pico/src/main.rs:568]:
```suggestion
        date_time,
```

File: boards/raspberry_pi_pico/src/main.rs [hunk 26]
```diff
@@ -551,7 +577,7 @@ pub unsafe fn main() {
     };
 
     debug!(
-        "RP2040 Revision {} {}",
+        "RP2040 Revision {} {}\n",
```
Comment (bradjc) [inline boards/raspberry_pi_pico/src/main.rs:580]:
```suggestion
        "RP2040 Revision {} {}",
```

File: capsules/extra/src/date_time.rs [hunk 27]
```diff
@@ -0,0 +1,354 @@
...
+            4 => Ok(date_time::DayOfWeek::Thursday),
+            5 => Ok(date_time::DayOfWeek::Friday),
+            6 => Ok(date_time::DayOfWeek::Saturday),
+            _ => Err(ErrorCode::INVAL),
+        }
+    }
+
+    /// Transforms two u32 numbers into a DateTimeValues structure (year, month, dotm, dotw, hour, minute, seconds)
+    /// Check file documentation for details on how the u32 tuple stores data
+
```
Comment (bradjc) [inline capsules/extra/src/date_time.rs:138]:
```suggestion
```

File: capsules/extra/src/date_time.rs [hunk 28]
```diff
@@ -0,0 +1,354 @@
...
+            return Err(ErrorCode::INVAL);
+        }
+
+        Ok(date_result)
+    }
+
+    /// Transforms DateTimeValues structure (year, month, dotm, dotw, hour, minute, seconds) into two u32 numbers
+    /// Check file documentation for details on how the u32 numbers stores data
+    /// The two u32 numbers are returned as a tuple
+
```
Comment (bradjc) [inline capsules/extra/src/date_time.rs:180]:
```suggestion
```

File: capsules/extra/src/date_time.rs (hunk 11 above)
Comment (bradjc) [reply capsules/extra/src/date_time.rs:279]: The code doesn't seem to track which event the app subscribed to.

File: chips/stm32f429zi/src/lib.rs [hunk 29]
```diff
@@ -15,6 +15,9 @@ pub mod interrupt_service;
 pub mod stm32f429zi_nvic;
 pub mod trng_registers;
 
+pub mod rtc;
+// pub mod pwr;
```
Comment (bradjc) [inline chips/stm32f429zi/src/lib.rs:19]:
```suggestion
```

File: boards/nucleo_f429zi/openocd.cfg (hunk 17 above)
Comment (Remus7) [reply boards/nucleo_f429zi/openocd.cfg:15]: Revert to what? As far as I remember, I never edited this file

Comment (bradjc) [reply boards/nucleo_f429zi/openocd.cfg:15]: If github is showing a change, then the file changed. Looks like a line was removed.

File: chips/stm32f429zi/src/pwr.rs [hunk 30]
```diff
@@ -0,0 +1,97 @@
...
+    /// Low-power deep sleep
+    LPDS OFFSET(0) NUMBITS(1) [],
+    /// Power down deepsleep
+    PDDS OFFSET(1) NUMBITS(1) [],
+    /// Clear wakeup flag
+    CWUF OFFSET(2) NUMBITS(1) [],
+    /// Clear standby flag
+    CSBF OFFSET(3) NUMBITS(1) [],
+    /// Power voltage detector
+/// enable
```
Comment (bradjc) [inline chips/stm32f429zi/src/pwr.rs:31]:
```suggestion
    /// enable
```

File: chips/stm32f429zi/src/pwr.rs [hunk 31]
```diff
@@ -0,0 +1,97 @@
...
+    PVDE OFFSET(4) NUMBITS(1) [],
+    /// PVD level selection
+    PLS OFFSET(5) NUMBITS(3) [],
+    /// Disable backup domain write
+/// protection
```
Comment (bradjc) [inline chips/stm32f429zi/src/pwr.rs:36]:
```suggestion
    /// protection
```

File: chips/stm32f429zi/src/pwr.rs [hunk 32]
```diff
@@ -0,0 +1,97 @@
...
+    DBP OFFSET(8) NUMBITS(1) [],
+    /// Flash power down in Stop
+/// mode
```
Comment (bradjc) [inline chips/stm32f429zi/src/pwr.rs:39]:
```suggestion
    /// mode
```

File: chips/stm32f429zi/src/pwr.rs [hunk 33]
```diff
@@ -0,0 +1,97 @@
...
+    FPDS OFFSET(9) NUMBITS(1) [],
+    /// Low-Power Regulator Low Voltage in
+/// deepsleep
```
Comment (bradjc) [inline chips/stm32f429zi/src/pwr.rs:42]:
```suggestion
    /// deepsleep
```

etc.

**Review by bradjc — COMMENTED** [review]
Very close!!

File: chips/stm32f429zi/src/pwr.rs
Comment (bradjc) [inline chips/stm32f429zi/src/pwr.rs:1]: Please fix the comment indentations.

**Comment by Remus7** [comment]
I deleted the endline in the openocd.cfg file, it seems that in my original text editor it was automatically ignored. The file should now be the same as in the main tock code.

**Review by bradjc — COMMENTED** [review]
I think this is close just need to remove extra changes and indent comments for register definitions to match.

**Comment by Remus7** [comment]
For some reason, after I ran make format to clear up indentation, the "Pull Request Labeler / triage" test fails with the message " YAMLException: can not read a block mapping entry; a multiline key may not be an implicit key (56:1)". Although I tried reverting the changes, the error persists. I do not have experience on this topic, would you have any suggestions as to how should I fix this issue?

**Comment by ppannuto** [comment]
Error in our tooling, sorry :(

Rebase/merge once https://github.com/tock/tock/pull/3673 goes through and it should be fixed.

File: chips/stm32f429zi/src/pwr.rs [hunk 34]
```diff
@@ -0,0 +1,97 @@
...
+/// selection
+    VOS OFFSET(14) NUMBITS(2) [
+        Scale3 = 0b01,
+        Scale2 = 0b10,
+        Scale1 = 0b11,
+    ],
+    /// Over-drive enable
+    ODEN OFFSET(16) NUMBITS(1) [],
+    /// Over-drive switching
+/// enabled
```
Comment (bradjc) [inline chips/stm32f429zi/src/pwr.rs:59]: indent

**Comment by bradjc** [comment]
Can you run `git checkout origin/master boards/nucleo_f429zi/openocd.cfg` and commit the result?

**Comment by Remus7** [comment]
Done.
Another thing; the ci-build and ci-tests from last time kept telling me that the "extract()" method is not found in `OptionalCell<DeferredCallTask>', although it seems to be there (and it compiles locally). Moreover, the test only complains about the method in the raspberry pi rtc, even though I use the exact same method in the nucleo board rtc. Should I delete it?

**Comment by bradjc** [comment]
The best thing to do would be to squash commits and then rebase on the latest master.

**Comment by Remus7** [comment]
How should I do that?

**Comment by Remus7** [comment]
I squashed all the commits, now waiting for the checks to finish.

File: capsules/extra/src/date_time.rs (hunk 27 above)
Comment (bradjc) [reply capsules/extra/src/date_time.rs:138]: Please remove this empty line.

File: capsules/extra/src/date_time.rs (hunk 28 above)
Comment (bradjc) [reply capsules/extra/src/date_time.rs:180]: Why do these keep getting resolved? The empty line is here and needs to be removed.

File: capsules/extra/src/date_time.rs (hunk 10 above)
Comment (bradjc) [reply capsules/extra/src/date_time.rs:259]: Wait how does this work? If an app calls get time while a different app calls set time, how does the second app actually get the time?

**Comment by Remus7** [comment]
The way it is meant to work, anytime an app calls to the capsule, its command is registered as subscribed. When the capsule ends the current task, it checks for any apps that have subscribed for a get or set operation. When a get operation is finished, the same result is passed to all apps that have subscribed for a get set (as they are the same values). Afterwards one of the queued set operations will be started. And then anther get operation.

**Comment by bradjc** [comment]
What happens in the case that:

1. App1 calls get()
2. While the get is being gotten, App2 calls set()

Won't App2 be issued a callback once the get finishes? Then App2 will continue execution as though the time was set, even though the set hasn't completed? Worse, the set will never actually happen, right?

**Comment by Remus7** [comment]
I seem to have uploaded an older, unfinished version of the capsule when I solved the rebase. Changing it now.

**Review by bradjc — APPROVED** [review]

**Comment by Remus7** [comment]
Would you happen to know why 5 of the tests take still haven't reported any status?

**Review by bradjc — DISMISSED** [review]

**Comment by bradjc** [comment]
Need to fix CI errors.

**Review by bradjc — APPROVED** [review]
